From 2bba307becccd085b624beab68f98486407799c4 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Qian=20Hu=20=28=E8=83=A1=E9=AA=9E=29?=
 <qian.hu@mediatek.com>
Date: Thu, 27 Nov 2025 15:32:24 +0800
Subject: [PATCH 06/14] v4l2object: providing aligned size when propose
 allocation

for multiplane, we also calculate aligned size. otherwise size = 0, and
upstream gst_v4l2_object_match_buffer_layout_from_struct will fail.

and, since we pass driver align size for upstream element, upstream element
should not just return when this size not match with calculated one

Upstream-Status: Submitted [https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/10214 ]

Signed-off-by: Raja Ganapathi Busam <rbusam@qti.qualcomm.com>
---
 sys/v4l2/gstv4l2object.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index 1730eb2..3c6db88 100644
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -5688,7 +5688,6 @@ gst_v4l2_object_match_buffer_layout_from_struct (GstV4l2Object * obj,
     GST_WARNING_OBJECT (obj->dbg_obj,
         "Requested buffer size (%d) doesn't match video info size (%"
         G_GSIZE_FORMAT ")", buffer_size, GST_VIDEO_INFO_SIZE (&info));
-    return FALSE;
   }
 
   GST_DEBUG_OBJECT (obj->dbg_obj,
@@ -6060,7 +6059,8 @@ gst_v4l2_object_propose_allocation (GstV4l2Object * obj, GstQuery * query)
    * between aligned size and actual size as extra padding to the upstream
    * element */
   if (V4L2_TYPE_IS_MULTIPLANAR (obj->type)) {
-    /* FIXME */
+    for (guint i = 0; i < obj->format.fmt.pix_mp.num_planes; i++)
+      aligned_size += obj->format.fmt.pix_mp.plane_fmt[i].sizeimage;
   } else {
     aligned_size = obj->format.fmt.pix.sizeimage;
   }
-- 
2.34.1

